File: Test 01/wave1d.cpp
#include <iostream>
#include <vector>
#include <cmath>
#include <fstream>
#include <chrono>
#include <string>
#include <iomanip>

int main() {
    using namespace std;
    using namespace std::chrono;

    // Parámetros físicos y malla
    const double L = 100.0;
    const double c = 10.0;
    const int m = 100;
    const double h = L / m;
    const double Lambda = 0.8;
    const double k = Lambda * h / c;
    const double T = 15.0;
    const int nt = static_cast<int>(ceil(T / k));
    const int n_inner = m - 1;

    vector<double> x(m + 1);
    for (int i = 0; i <= m; ++i)
        x[i] = i * h;

    // Condiciones iniciales
    auto f_init = [&](double xx) { return sin(M_PI * xx / L); };    // f(x) = u(x,0) -> desplazamiento inicial
    auto g_init = [&](double) { return 0.0; };                     // g(x) = u_t(x,0) -> velocidad inicial

    vector<double> u0(n_inner), v0(n_inner);
    for (int i = 0; i < n_inner; ++i) {
        u0[i] = f_init(x[i + 1]);
        v0[i] = g_init(x[i + 1]);
    }

    // Construcción de matriz M
    vector<vector<double>> Mmat(n_inner, vector<double>(n_inner, 0.0));
    for (int i = 0; i < n_inner; ++i)
        Mmat[i][i] = 2.0 * (1.0 - Lambda * Lambda);
    for (int i = 0; i < n_inner - 1; ++i) {
        Mmat[i][i + 1] += Lambda * Lambda;
        Mmat[i + 1][i] += Lambda * Lambda;
    }

    // Inicialización temporal
    vector<double> u_prev = u0;          // u^(0)
    vector<double> u_curr(n_inner, 0.0); // u^(1)

    // Primer paso en tiempo
    for (int i = 0; i < n_inner; ++i)
        u_curr[i] = (1.0 - Lambda * Lambda) * u0[i] + k * v0[i];

    for (int i = 1; i < n_inner - 1; ++i)
        u_curr[i] += (Lambda * Lambda / 2.0) * (u0[i - 1] + u0[i + 1]);

    // aplicar condiciones de frontera (u=0 en extremos)
    u_curr[0] = (1.0 - Lambda * Lambda) * u0[0]
                + (Lambda * Lambda / 2.0) * (u0[1])
                + k * v0[0];

    u_curr[n_inner - 1] = (1.0 - Lambda * Lambda) * u0[n_inner - 1]
                          + (Lambda * Lambda / 2.0) * (u0[n_inner - 2])
                          + k * v0[n_inner - 1];

    // Evolución temporal
    vector<vector<double>> snapshots;
    vector<double> times;

    // Guardar primeros estados
    vector<double> full0(m + 1, 0.0);
    for (int i = 1; i < m; ++i) full0[i] = u_prev[i - 1];
    snapshots.push_back(full0);
    times.push_back(0.0);

    vector<double> full1(m + 1, 0.0);
    for (int i = 1; i < m; ++i) full1[i] = u_curr[i - 1];
    snapshots.push_back(full1);
    times.push_back(k);

    vector<double> u_next(n_inner, 0.0);

    cout << "Iniciando integración temporal..." << endl;
    auto t0 = high_resolution_clock::now();

    for (int j = 1; j < nt; ++j) {
        // u^{j+1} = M * u^{j} - u^{j-1}
        for (int i = 0; i < n_inner; ++i) {
            double sum = 0.0;
            for (int l = 0; l < n_inner; ++l)
                sum += Mmat[i][l] * u_curr[l];
            u_next[i] = sum - u_prev[i];
        }

        // Guardar snapshots cada cierto número de pasos
        if (j % max(1, nt / 200) == 0) {
            vector<double> full(m + 1, 0.0);
            for (int i = 1; i < m; ++i)
                full[i] = u_next[i - 1];
            snapshots.push_back(full);
            times.push_back((j + 1) * k);
        }
        u_prev = u_curr;
        u_curr = u_next;
    }

    auto t1 = high_resolution_clock::now();
    double elapsed = duration<double>(t1 - t0).count();
    cout << "Integración completada en " << elapsed << " s, "
         << snapshots.size() << " snapshots almacenados." << endl;

    // Exportar a CSV
    string filename = "wave_cpp.csv";
    ofstream csv(filename);
    csv << fixed << setprecision(6);

    // encabezado
    csv << "t";
    for (int i = 0; i <= m; ++i)
        csv << ",x" << i;
    csv << "\n";

    // datos
    for (size_t j = 0; j < snapshots.size(); ++j) {
        csv << times[j];
        for (int i = 0; i <= m; ++i)
            csv << "," << snapshots[j][i];
        csv << "\n";
    }

    csv.close();
    return 0;
}

File: Test 01/wave1d.f90
! wave1d_fortran.f90
program wave1d_fortran
  implicit none

  ! ---------- precision ----------
  integer, parameter :: dp = selected_real_kind(15, 307)

  ! ---------- parámetros del problema ----------
  real(dp), parameter :: Ldom = 100.0_dp     ! longitud del dominio
  real(dp), parameter :: c = 10.0_dp         ! velocidad de la onda
  integer, parameter :: m = 100              ! número de subintervalos
  real(dp), parameter :: h = Ldom / real(m, dp) ! paso espacial
  real(dp), parameter :: Lambda = 0.8_dp     ! número de Courant
  real(dp), parameter :: kstep = Lambda * h / c  ! paso temporal
  real(dp), parameter :: T = 15.0_dp         ! tiempo final
  
  ! Constantes matemáticas
  real(dp), parameter :: pi = 4.0_dp * atan(1.0_dp)
  
  ! Variables derivadas
  integer :: nt, n_inner, save_interval, n_snap_expected
  integer :: ii, jj, ll, sn, unit_bin, unit_meta, ios
  real(dp) :: tstart, tend, wall_time, sum_val
  character(len=128) :: filename_bin, filename_meta
  
  ! ---------- arrays allocatables ----------
  real(dp), allocatable :: x(:)              ! puntos de la malla 0..m
  real(dp), allocatable :: u0(:), v0(:)      ! condiciones iniciales
  real(dp), allocatable :: u_prev(:)         ! u^(j-1)
  real(dp), allocatable :: u_curr(:)         ! u^(j)  
  real(dp), allocatable :: u_next(:)         ! u^(j+1)
  real(dp), allocatable :: Mmat(:,:)         ! matriz M
  real(dp), allocatable :: snapshots(:,:)    ! matriz de snapshots
  real(dp), allocatable :: times(:)          ! tiempos de los snapshots

  ! ---------- cálculo de dimensiones ----------
  nt = ceiling(T / kstep)
  n_inner = m - 1
  save_interval = max(1, nt / 200)
  n_snap_expected = 2 + (nt - 1) / save_interval

  ! ---------- asignación de memoria ----------
  allocate(x(0:m))
  allocate(u0(n_inner), v0(n_inner))
  allocate(u_prev(n_inner), u_curr(n_inner), u_next(n_inner))
  allocate(Mmat(n_inner, n_inner))
  allocate(snapshots(n_snap_expected, 0:m))
  allocate(times(n_snap_expected))

  ! ---------- inicialización de la malla ----------
  do ii = 0, m
    x(ii) = real(ii, dp) * h
  end do

  ! ---------- condiciones iniciales ----------
  do ii = 1, n_inner
    u0(ii) = sin(pi * x(ii) / Ldom)    ! desplazamiento inicial f(x)
    v0(ii) = 0.0_dp                    ! velocidad inicial g(x) = 0
  end do

  ! ---------- construcción de la matriz M ----------
  ! La matriz M implementa: u^(n+1) = M * u^(n) - u^(n-1)
  ! Donde M * u^(n) = 2u^n - λ²(u^n_{i-1} - 2u^n_i + u^n_{i+1})
  !                 = λ²u^n_{i-1} + (2 - 2λ²)u^n_i + λ²u^n_{i+1}
  
  Mmat = 0.0_dp
  
  ! Diagonal principal: 2 - 2λ²
  do ii = 1, n_inner
    Mmat(ii, ii) = 2.0_dp - 2.0_dp * Lambda * Lambda
  end do
  
  ! Subdiagonales: λ²
  do ii = 1, n_inner - 1
    Mmat(ii, ii + 1) = Lambda * Lambda
    Mmat(ii + 1, ii) = Lambda * Lambda
  end do

  ! ---------- inicialización temporal ----------
  u_prev = u0  ! u^(0)
  
  ! Cálculo de u^(1) usando el esquema de Taylor para el primer paso
  ! u^1 = u^0 + k*v^0 + (k²/2)*u_tt^0
  ! donde u_tt = c²*u_xx, entonces: u^1 = u^0 + k*v^0 + (λ²/2)*(u^0_{i-1} - 2u^0_i + u^0_{i+1})
  do ii = 1, n_inner
    if (ii == 1) then
      ! Punto adyacente a frontera izquierda (u0(0) = 0)
      u_curr(ii) = u0(ii) + kstep * v0(ii) + &
                   (Lambda * Lambda / 2.0_dp) * (0.0_dp - 2.0_dp * u0(ii) + u0(ii + 1))
    else if (ii == n_inner) then
      ! Punto adyacente a frontera derecha (u0(n_inner+1) = 0)
      u_curr(ii) = u0(ii) + kstep * v0(ii) + &
                   (Lambda * Lambda / 2.0_dp) * (u0(ii - 1) - 2.0_dp * u0(ii) + 0.0_dp)
    else
      ! Puntos internos
      u_curr(ii) = u0(ii) + kstep * v0(ii) + &
                   (Lambda * Lambda / 2.0_dp) * (u0(ii - 1) - 2.0_dp * u0(ii) + u0(ii + 1))
    end if
  end do

  ! ---------- almacenamiento de snapshots iniciales ----------
  sn = 1
  
  ! Snapshot en t = 0
  snapshots(sn, 0) = 0.0_dp      ! frontera izquierda
  do ii = 1, n_inner
    snapshots(sn, ii) = u_prev(ii) ! puntos internos
  end do
  snapshots(sn, m) = 0.0_dp      ! frontera derecha
  times(sn) = 0.0_dp
  
  ! Snapshot en t = kstep
  sn = sn + 1
  snapshots(sn, 0) = 0.0_dp      ! frontera izquierda
  do ii = 1, n_inner
    snapshots(sn, ii) = u_curr(ii) ! puntos internos
  end do
  snapshots(sn, m) = 0.0_dp      ! frontera derecha
  times(sn) = kstep
  sn = sn + 1

  ! ---------- integración temporal ----------
  call cpu_time(tstart)

  do jj = 1, nt - 1
    ! u^{j+1} = M * u^{j} - u^{j-1}
    do ii = 1, n_inner
      sum_val = 0.0_dp
      do ll = 1, n_inner
        sum_val = sum_val + Mmat(ii, ll) * u_curr(ll)
      end do
      u_next(ii) = sum_val - u_prev(ii)
    end do

    ! Guardar snapshot periódicamente
    if (mod(jj, save_interval) == 0) then
      snapshots(sn, 0) = 0.0_dp      ! frontera izquierda
      do ii = 1, n_inner
        snapshots(sn, ii) = u_next(ii) ! puntos internos
      end do
      snapshots(sn, m) = 0.0_dp      ! frontera derecha
      times(sn) = real(jj, dp) * kstep
      sn = sn + 1
    end if

    ! Avanzar en el tiempo
    u_prev = u_curr
    u_curr = u_next
  end do

  call cpu_time(tend)
  wall_time = tend - tstart

  ! Ajustar número real de snapshots
  n_snap_expected = sn - 1

  write(*, '(A, F8.4, A, I0, A)') 'Integración completada en ', wall_time, &
       ' s, ', n_snap_expected, ' snapshots almacenados.'

  ! ---------- exportar datos binarios ----------
  filename_bin = 'wave1d_fortran.bin'
  filename_meta = 'wave1d_meta.txt'
  
  ! Escribir metadata
  open(newunit=unit_meta, file=filename_meta, status='replace', action='write', iostat=ios)
  if (ios == 0) then
    write(unit_meta, *) n_snap_expected, m + 1
    close(unit_meta)
  else
    write(*,*) 'Error escribiendo archivo de metadata'
  end if

  ! Escribir datos binarios
  open(newunit=unit_bin, file=filename_bin, status='replace', action='write', &
       form='unformatted', access='stream', iostat=ios)
  if (ios == 0) then
    write(unit_bin) snapshots(1:n_snap_expected, 0:m)
    close(unit_bin)
    write(*,*) 'Archivos binarios generados: ', trim(filename_bin), ' y ', trim(filename_meta)
  else
    write(*,*) 'Error escribiendo archivo binario'
  end if

  ! ---------- liberar memoria ----------
  deallocate(x, u0, v0, u_prev, u_curr, u_next, Mmat, snapshots, times)

end program wave1d_fortran